File: src/main/java/org/almibe/multipage/PageBuilder.java
package org.almibe.multipage;

import javafx.beans.property.ReadOnlyStringProperty;
import javafx.scene.Node;
import javafx.scene.input.KeyCombination;

import java.util.Map;
import java.util.concurrent.Callable;

public class PageBuilder {
    private Node content;
    private ReadOnlyStringProperty title;
    private Node icon;
    private Callable<Boolean> allowClose;
    private Map<KeyCombination, Runnable> accelerators;

    public PageBuilder setContent(Node content) {
        this.content = content;
        return this;
    }

    public PageBuilder setTitle(ReadOnlyStringProperty title) {
        this.title = title;
        return this;
    }

    public PageBuilder setIcon(Node icon) {
        this.icon = icon;
        return this;
    }

    public PageBuilder setAllowClose(Callable<Boolean> allowClose) {
        this.allowClose = allowClose;
        return this;
    }

    public PageBuilder setAccelerators(Map<KeyCombination, Runnable> accelerators) {
        this.accelerators = accelerators;
        return this;
    }

    public Page createPage() {
        return new Page() {
            @Override public Node content() {
                return content;
            }
            @Override public ReadOnlyStringProperty title() {
                return title;
            }
            @Override public Node icon() {
                return icon;
            }
            @Override public Callable<Boolean> allowClose() {
                return allowClose;
            }
            @Override public Map<KeyCombination, Runnable> accelerators() {
                return accelerators;
            }
        };
    }
}

File: src/main/kotlin/org/almibe/multipage/PageBuilder.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.almibe.multipage

import java.awt.KeyEventDispatcher
import javax.swing.ImageIcon
import javax.swing.JComponent

class PageBuilder {
    private var content: JComponent? = null
    private var title: String? = null
    private var icon: ImageIcon? = null
    private var allowClose: (() -> Boolean)? = null
    private var keyEventDispatcher: KeyEventDispatcher? = null

    fun setContent(content: JComponent): PageBuilder {
        this.content = content
        return this
    }

    fun setTitle(title: String): PageBuilder {
        this.title = title
        return this
    }

    fun setIcon(icon: ImageIcon): PageBuilder {
        this.icon = icon
        return this
    }

    fun setAllowClose(allowClose: () -> Boolean): PageBuilder {
        this.allowClose = allowClose
        return this
    }

    fun setKeyEventDispatcher(keyEventDispatcher: KeyEventDispatcher): PageBuilder {
        this.keyEventDispatcher = keyEventDispatcher
        return this
    }

    fun createPage(): Page {
        return Page(content!!, title!!, icon, allowClose, keyEventDispatcher)
    }
}
